File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/model/SocProfile.java
package uk.gov.justice.digital.hmpps.riskprofiler.model;

import io.swagger.annotations.ApiModelProperty;
import lombok.*;
import org.apache.commons.lang3.builder.CompareToBuilder;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;

@Data
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = true)
public class SocProfile extends RiskProfile implements Comparable<SocProfile> {
    @ApiModelProperty(value = "Indicates the offender must be transferred to security", example = "true", position = 4)
    private boolean transferToSecurity;

    public RiskType getRiskType() {
        return RiskType.SOC;
    }

    @Builder(builderMethodName = "socBuilder")
    public SocProfile(@NotBlank String nomsId, @NotBlank String provisionalCategorisation, boolean transferToSecurity) {
        super(nomsId, provisionalCategorisation);
        this.transferToSecurity = transferToSecurity;
    }

    public int compareTo(@NotNull SocProfile socProfile) {
        return new CompareToBuilder()
                        .append(this.getRiskType(), socProfile.getRiskType())
                        .append(this.getNomsId(), socProfile.getNomsId())
                        .append(socProfile.isTransferToSecurity(), this.isTransferToSecurity())
                        .append(this.getProvisionalCategorisation(), socProfile.getProvisionalCategorisation())
                        .toComparison();
    }

}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/model/SocProfile.kt
package uk.gov.justice.digital.hmpps.riskprofiler.model

import io.swagger.v3.oas.annotations.media.Schema
import org.apache.commons.lang3.builder.CompareToBuilder
import javax.validation.constraints.NotBlank

data class SocProfile(
  val nomsId: @NotBlank String,
  var provisionalCategorisation: @NotBlank String,

  @field:Schema(
    title = "Indicates the offender must be transferred to security",
    example = "true",
  )
  var transferToSecurity: Boolean,
) : RiskProfile(nomsId, provisionalCategorisation), Comparable<SocProfile> {

  constructor() : this("", "", false)

  override fun getRiskType(): RiskType {
    return RiskType.SOC
  }

  override fun compareTo(other: SocProfile): Int {
    return CompareToBuilder()
      .append(getRiskType(), other.getRiskType())
      .append(this.nomsId, other.nomsId)
      .append(other.transferToSecurity, this.transferToSecurity)
      .append(this.provisionalCategorisation, other.provisionalCategorisation)
      .toComparison()
  }
}
